File: bets/f-cn-work-agent/mock-webhook-receiver.py
#!/usr/bin/env python3
"""Tiny mock HTTP webhook receiver for F cn-work-agent local-mvp.

Writes the last POST body to --out (default data/webhook-last.json).
Optional --secret: verify X-Webhook-Signature HMAC-SHA256 of the raw body.
Optional --headers-out: persist last request headers (+ verified flag + timestamp).
Records X-Webhook-Timestamp when present (OSS; replay window = paid).

  python3 mock-webhook-receiver.py --port 8810 --out data/webhook-last.json
  python3 mock-webhook-receiver.py --port 8812 --secret whsec_local_mvp \
    --out data/webhook-hmac-last.json --headers-out data/webhook-hmac-last.headers.json
"""
from __future__ import annotations

import argparse
import json
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import urlparse

# Allow running from bet root without PYTHONPATH.
sys.path.insert(0, str(Path(__file__).resolve().parent / "src"))
from cn_work_agent.webhook import SIGNATURE_HEADER, TIMESTAMP_HEADER, verify_webhook_signature  # noqa: E402


def main() -> int:
    parser = argparse.ArgumentParser(prog="mock-webhook-receiver")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=8810)
    parser.add_argument("--out", default="data/webhook-last.json")
    parser.add_argument(
        "--headers-out",
        default=None,
        help="Write last request headers JSON (signature + verified). "
        "Default: <out>.headers.json when --secret is set.",
    )
    parser.add_argument(
        "--secret",
        default=None,
        help="HMAC-SHA256 key; verify X-Webhook-Signature of the raw body. "
        "Empty/omit = accept unsigned.",
    )
    args = parser.parse_args()
    out = Path(args.out)
    if not out.is_absolute():
        out = Path.cwd() / out
    out.parent.mkdir(parents=True, exist_ok=True)

    secret = str(args.secret).strip() if args.secret else None
    secret = secret or None

    headers_out = args.headers_out
    if headers_out:
        headers_abs = Path(headers_out)
        if not headers_abs.is_absolute():
            headers_abs = Path.cwd() / headers_abs
    elif secret:
        name = out.name
        if name.endswith(".json"):
            headers_abs = out.with_name(name[: -len(".json")] + ".headers.json")
        else:
            headers_abs = out.with_name(name + ".headers.json")
    else:
        headers_abs = None
    if headers_abs is not None:
        headers_abs.parent.mkdir(parents=True, exist_ok=True)

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *a) -> None:  # noqa: A002
            print("%s - %s" % (self.address_string(), fmt % a), flush=True)

        def do_GET(self) -> None:  # noqa: N802
            path = urlparse(self.path).path
            if path == "/health":
                body = json.dumps({"ok": True, "service": "mock-webhook-receiver"}).encode(
                    "utf-8"
                )
                self.send_response(200)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Content-Length", str(len(body)))
                self.end_headers()
                self.wfile.write(body)
                return
            self.send_response(404)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            body = b'{"error":"not_found"}'
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

        def do_POST(self) -> None:  # noqa: N802
            length = int(self.headers.get("Content-Length") or "0")
            raw = self.rfile.read(length) if length else b""
            sig_header = self.headers.get(SIGNATURE_HEADER) or ""
            ts_header = self.headers.get(TIMESTAMP_HEADER) or ""
            verified = None
            if secret:
                verified = verify_webhook_signature(secret, raw, sig_header)
            try:
                out.write_bytes(raw if raw else b"")
                if headers_abs is not None:
                    meta = {
                        "signature": sig_header or None,
                        "timestamp": ts_header or None,
                        "verified": verified,
                        "headers": {k: v for k, v in self.headers.items()},
                    }
                    headers_abs.write_text(
                        json.dumps(meta, indent=2) + "\n", encoding="utf-8"
                    )
            except OSError as e:
                msg = json.dumps({"ok": False, "error": str(e)}).encode("utf-8")
                self.send_response(500)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Content-Length", str(len(msg)))
                self.end_headers()
                self.wfile.write(msg)
                return
            if secret and verified is False:
                deny = json.dumps(
                    {
                        "ok": False,
                        "error": "invalid_signature",
                        "received": True,
                        "verified": False,
                        "bytes": len(raw),
                    }
                ).encode("utf-8")
                self.send_response(401)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Content-Length", str(len(deny)))
                self.end_headers()
                self.wfile.write(deny)
                return
            ack = json.dumps(
                {
                    "ok": True,
                    "received": True,
                    "bytes": len(raw),
                    "verified": verified,
                }
            ).encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            self.send_header("Content-Length", str(len(ack)))
            self.end_headers()
            self.wfile.write(ack)

    server = ThreadingHTTPServer((args.host, args.port), Handler)
    extra = ""
    if secret:
        extra += " verify=hmac"
    if headers_abs is not None:
        extra += f" headers={headers_abs}"
    print(
        f"mock-webhook-receiver listening on http://{args.host}:{args.port} out={out}{extra}",
        flush=True,
    )
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nshutting down", flush=True)
    finally:
        server.server_close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bets/b-mcp-gateway/mock-webhook-receiver.js
#!/usr/bin/env node
/**
 * Tiny mock HTTP webhook receiver for B gateway local-mvp.
 * Writes the last POST body to --out (default data/webhook-last.json).
 * Optional --secret: verify X-Webhook-Signature HMAC-SHA256 of the raw body.
 * Optional --headers-out: persist last request headers (+ verified flag + timestamp).
 * Records X-Webhook-Timestamp when present (OSS; replay window = paid).
 * Optional --fail-once: first POST returns 500 (body not persisted); later POSTs 200.
 * GET /stats → {ok, requests} (POST count).
 *
 *   node mock-webhook-receiver.js --port 8792 --out data/webhook-last.json
 *   node mock-webhook-receiver.js --port 8797 --secret whsec_local_mvp \
 *     --out data/webhook-hmac-last.json --headers-out data/webhook-hmac-last.headers.json
 *   node mock-webhook-receiver.js --port 8785 --fail-once --out data/webhook-retry-last.json
 */
import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import { verifyWebhookSignature, SIGNATURE_HEADER, TIMESTAMP_HEADER } from "./src/webhooks.js";

function parseArgs(argv) {
  let port = 8792;
  let host = "127.0.0.1";
  let out = "data/webhook-last.json";
  let headersOut = null;
  let secret = null;
  let failOnce = false;
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--port") port = Number(argv[++i]);
    else if (a === "--host") host = argv[++i];
    else if (a === "--out") out = argv[++i];
    else if (a === "--headers-out") headersOut = argv[++i];
    else if (a === "--secret") secret = argv[++i];
    else if (a === "--fail-once") failOnce = true;
  }
  if (typeof secret === "string") {
    secret = secret.trim() || null;
  } else {
    secret = null;
  }
  return { port, host, out, headersOut, secret, failOnce };
}

const { port, host, out, headersOut, secret, failOnce } = parseArgs(process.argv);
const outAbs = path.isAbsolute(out) ? out : path.resolve(process.cwd(), out);
const headersAbs = headersOut
  ? path.isAbsolute(headersOut)
    ? headersOut
    : path.resolve(process.cwd(), headersOut)
  : secret
    ? outAbs.replace(/(\.json)?$/, ".headers.json")
    : null;

function writeFileSafe(abs, contents) {
  fs.mkdirSync(path.dirname(abs), { recursive: true });
  fs.writeFileSync(abs, contents, "utf8");
}

let postCount = 0;
let failRemaining = failOnce ? 1 : 0;

const server = http.createServer((req, res) => {
  const url = new URL(req.url || "/", `http://${req.headers.host || "localhost"}`);
  const method = (req.method || "GET").toUpperCase();

  if (method === "GET" && url.pathname === "/health") {
    const body = JSON.stringify({ ok: true, service: "mock-webhook-receiver" });
    res.writeHead(200, {
      "content-type": "application/json; charset=utf-8",
      "content-length": Buffer.byteLength(body),
    });
    return res.end(body);
  }

  if (method === "GET" && url.pathname === "/stats") {
    const body = JSON.stringify({ ok: true, requests: postCount });
    res.writeHead(200, {
      "content-type": "application/json; charset=utf-8",
      "content-length": Buffer.byteLength(body),
    });
    return res.end(body);
  }

  if (method === "POST") {
    const chunks = [];
    req.on("data", (c) => chunks.push(c));
    req.on("end", () => {
      postCount += 1;
      const raw = Buffer.concat(chunks).toString("utf8");
      if (failRemaining > 0) {
        failRemaining -= 1;
        const fail = JSON.stringify({ ok: false, error: "fail_once", requests: postCount });
        res.writeHead(500, {
          "content-type": "application/json; charset=utf-8",
          "content-length": Buffer.byteLength(fail),
        });
        return res.end(fail);
      }
      const sigHeader = req.headers["x-webhook-signature"] || req.headers[SIGNATURE_HEADER.toLowerCase()] || "";
      const tsHeader = req.headers["x-webhook-timestamp"] || req.headers[TIMESTAMP_HEADER.toLowerCase()] || "";
      let verified = null;
      if (secret) {
        verified = verifyWebhookSignature(secret, raw, sigHeader);
      }
      try {
        writeFileSafe(outAbs, raw || "");
        if (headersAbs) {
          const meta = {
            signature: sigHeader || null,
            timestamp: tsHeader || null,
            verified,
            headers: { ...req.headers },
          };
          writeFileSafe(headersAbs, JSON.stringify(meta, null, 2) + "\n");
        }
      } catch (err) {
        const msg = JSON.stringify({ ok: false, error: String(err?.message || err) });
        res.writeHead(500, {
          "content-type": "application/json; charset=utf-8",
          "content-length": Buffer.byteLength(msg),
        });
        return res.end(msg);
      }
      if (secret && verified === false) {
        const deny = JSON.stringify({
          ok: false,
          error: "invalid_signature",
          received: true,
          verified: false,
          bytes: Buffer.byteLength(raw),
        });
        res.writeHead(401, {
          "content-type": "application/json; charset=utf-8",
          "content-length": Buffer.byteLength(deny),
        });
        return res.end(deny);
      }
      const ack = JSON.stringify({
        ok: true,
        received: true,
        bytes: Buffer.byteLength(raw),
        verified,
      });
      res.writeHead(200, {
        "content-type": "application/json; charset=utf-8",
        "content-length": Buffer.byteLength(ack),
      });
      res.end(ack);
    });
    req.on("error", () => {
      res.writeHead(400);
      res.end();
    });
    return;
  }

  res.writeHead(404, { "content-type": "application/json; charset=utf-8" });
  res.end(JSON.stringify({ error: "not_found" }));
});

server.listen(port, host, () => {
  console.log(
    `mock-webhook-receiver listening on http://${host}:${port} out=${outAbs}` +
      (secret ? " verify=hmac" : "") +
      (failOnce ? " fail-once" : "") +
      (headersAbs ? ` headers=${headersAbs}` : "")
  );
});
